packages/react-keytips/src/components/Keytips/useKeytips.tsx

```diff
@@ -0,0 +1,261 @@
+    };
+  }, []);
+
+  React.useEffect(() => {
+    const handleDismiss = (ev: Event) => {
+      if (inKeytipMode) {
+        handleExitKeytipMode(ev as KeyboardEvent);
+      }
+    };
+
+    const addEventListeners = () => {
+      targetDocument?.addEventListener('mouseup', handleDismiss);
+      targetDocument?.defaultView?.addEventListener('resize', handleDismiss);
+      targetDocument?.defaultView?.addEventListener('scroll', handleDismiss);
+    };
+
+    const removeEventListeners = () => {
+      targetDocument?.removeEventListener('mouseup', handleDismiss);
+      targetDocument?.defaultView?.removeEventListener('resize', handleDismiss);
+      targetDocument?.defaultView?.removeEventListener('scroll', handleDismiss);
+    };
+
+    addEventListeners();
+    return removeEventListeners;
+  }, [targetDocument, inKeytipMode, handleExitKeytipMode]);
+
+  React.useEffect(() => {
+    const handleKeyDown = (ev: KeyboardEvent) => {
+      ev.preventDefault();
+      ev.stopPropagation();
+
+      if (inKeytipMode) {
+        const { key } = parseHotkey(ev.key.toLowerCase());
+        const currSeq = currentSequence.current + key?.toLowerCase();
+        const node = tree.getMatchingNode(currSeq);
+
+        if (node) {
+          tree.currentKeytip.current = node;
+          const currentKtpChildren = tree.getChildren(node);
+          if (node.target) {
+            node.onExecute?.(ev, {
+              event: ev,
+              type: 'keydown',
+              targetElement: node.target,
+            });
+          }
+          // To exit keytipMode after executing the keytip it must not have a menu or have dynamic children
+          if (currentKtpChildren.length === 0 && !node.dynamic) {
+            handleExitKeytipMode(ev);
+          } else {
+            // show all children keytips
+            showKeytips(currentKtpChildren);
+          }
+
+          currentSequence.current = '';
+          return;
+        }
+
+        // if we don't have a match, we have to check if the sequence is a partial match
+        const partialNodes = tree.getPartiallyMatched(currSeq);
+        if (partialNodes && partialNodes.length > 0) {
+          // We found nodes that partially match the sequence, so we show only those
+          const ids = partialNodes.map((n) => (n ? n.uniqueId : ''));
+          showKeytips(ids);
+          // Save currentSequence
+          currentSequence.current = currSeq;
+        }
+      }
+    };
+
+    targetDocument?.addEventListener('keydown', handleKeyDown);
+    return () => {
+      targetDocument?.removeEventListener('keydown', handleKeyDown);
+    };
+  }, [inKeytipMode, handleExitKeytipMode]);
+
+  const visibleKeytips = Object.entries(keytips)
+    .filter(([, { visible, visibleInternal }]) => visible || visibleInternal)
+    .map(([keytipId, keytipProps]) => (
+      <Keytip key={keytipId} {...keytipProps} />
+    ));
+
+  const hiddenKeytips = Object.values(keytips).map(({ keySequences }) => (
```

Comment:
this is similar to v8 keytips implementation and needed for maintaining ids for `aria-describedby`